lib/srv/server/ssm_install.go

```diff
@@ -0,0 +1,165 @@
+*/
+
+package server
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/gravitational/teleport/api/types/events"
+	libevent "github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/trace"
+)
+
+type SSMInstaller struct {
+	instances    []*string
+	SSM          ssmiface.SSMAPI
+	recheckTimer *time.Ticker
+	params       map[string][]*string
+	emitter      events.Emitter
+	ctx          context.Context
+	cancelfn     context.CancelFunc
+	region       string
+	accountID    string
+}
+
+// SSMInstallerConfig represents configuration for an SSM install
+// script executor.
+type SSMInstallerConfig struct {
+	// SSM is an SSM API client.
+	SSM ssmiface.SSMAPI
+	// Instances is the list of instances that will have the SSM
+	// document executed on them.
+	Instances []*ec2.Instance
+	// Params is a list of parameters to include when executing the
+	// SSM document.
+	Params map[string]string
+	// Emitter is an events emitter.
+	Emitter events.Emitter
+	Ctx     context.Context
+	// Region is the region instances are present in, used in audit
+	// events.
+	Region string
+	// AccountID is the AWS account being used to execute the SSM document.
+	AccountID string
+}
+
+func NewSSMInstaller(cfg SSMInstallerConfig) *SSMInstaller {
+	var ids []*string
+
+	for _, inst := range cfg.Instances {
+		ids = append(ids, inst.InstanceId)
+	}
+	ssmParams := make(map[string][]*string)
+
+	for key, val := range cfg.Params {
+		ssmParams[key] = []*string{aws.String(val)}
+	}
+
+	return &SSMInstaller{
+		instances:    ids,
+		SSM:          cfg.SSM,
+		recheckTimer: time.NewTicker(time.Second * 30),
+		params:       ssmParams,
+		emitter:      cfg.Emitter,
+		ctx:          cfg.Ctx,
+		region:       cfg.Region,
+		accountID:    cfg.AccountID,
+	}
+}
+
+var ErrCommandInProgress = errors.New("command in progress")
+
+func (i *SSMInstaller) checkCommands(commandID *string) error {
+	for _, inst := range i.instances {
+		cmdOut, err := i.SSM.GetCommandInvocation(&ssm.GetCommandInvocationInput{
+			CommandId:  commandID,
+			InstanceId: inst,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		status := aws.StringValue(cmdOut.Status)
+		if status == ssm.CommandStatusInProgress {
+			return trace.Wrap(ErrCommandInProgress)
+		}
+
+		var code string
+		if status == ssm.CommandStatusFailed {
+			code = libevent.SSMRunFailCode
+		} else {
+			code = libevent.SSMRunSuccessCode
+		}
+
+		event := events.SSMRun{
+			Metadata: events.Metadata{
+				Type: libevent.SSMRunEvent,
+				Code: code,
+			},
+			CommandID:  aws.StringValue(commandID),
+			InstanceID: aws.StringValue(inst),
+			AccountID:  i.accountID,
+			Region:     i.region,
+			ExitCode:   aws.Int64Value(cmdOut.ResponseCode),
+			Status:     status,
+		}
+
+		if err := i.emitter.EmitAuditEvent(i.ctx, &event); err != nil {
```

Comment:
The `checkCommands` function can run multiple times for the same set of servers, right? If e.g. some are still executing. Doesn't this mean that this will emit multiple audit events for the same server i.e. we'll get duplicates?